app/admin.go

```diff
@@ -56,16 +56,81 @@ func (s *Server) GetLogs(page, perPage int) ([]string, *model.AppError) {
 	return lines, nil
 }
 
+func (s *Server) QueryLogs(page, perPage int, logFilter *model.LogFilter) (map[string][]string, *model.AppError) {
+	logData := make(map[string][]string)
+
+	serverName := "default"
+
+	license := s.License()
+	if license != nil && *license.Features.Cluster && s.platform.Cluster() != nil && *s.platform.Config().ClusterSettings.Enable {
+		if info := s.platform.Cluster().GetMyClusterInfo(); info != nil {
+			serverName = info.Hostname
+		} else {
+			mlog.Error("Could not get cluster info")
+		}
+	}
+
+	serverNames := logFilter.ServerNames
+	if len(serverNames) > 0 {
+		for _, nodeName := range serverNames {
+			if nodeName == "default" {
+				AddLocalLogs(s, page, perPage, logData, nodeName, logFilter)
+			}
+		}
+	} else {
+		AddLocalLogs(s, page, perPage, logData, serverName, logFilter)
+	}
+
+	if s.platform.Cluster() != nil && *s.Config().ClusterSettings.Enable {
+		clusterLogs, err := s.platform.Cluster().QueryLogs(page, perPage)
+		if err != nil {
+			return nil, err
+		}
+
+		if clusterLogs != nil {
+			if len(serverNames) > 0 {
+				for _, filteredNodeName := range serverNames {
+					for nodeName, logs := range clusterLogs {
+						if nodeName == filteredNodeName {
+							logData[nodeName] = logs
+						}
+					}
+				}
+			} else {
+				for nodeName, logs := range clusterLogs {
+					logData[nodeName] = logs
+				}
+			}
+		}
+	}
+
+	return logData, nil
+}
+
+func AddLocalLogs(s *Server, page, perPage int, logData map[string][]string, serverName string, logFilter *model.LogFilter) *model.AppError {
```

Comment:
Can we make `logData` the first element in the args to indicate that it's the receiver? Also it might be a good idea to comment this. 
```suggestion
func AddLocalLogs( logData map[string][]string, s *Server, page, perPage int, serverName string, logFilter *model.LogFilter) *model.AppError {
```